src/agreement/mod.rs

```diff
@@ -467,75 +409,138 @@ impl<NodeUid: Clone + Debug + Ord> Agreement<NodeUid> {
         sender_id: &NodeUid,
         msg: CommonCoinMessage,
     ) -> Result<Step<NodeUid>> {
-        let coin_step = self.common_coin.handle_message(sender_id, msg)?;
+        let coin_step = match self.coin_state {
+            CoinState::Decided(_) => return Ok(Step::default()), // Coin value is already decided.
+            CoinState::InProgress(ref mut common_coin) => {
+                common_coin.handle_message(sender_id, msg)?
+            }
+        };
         self.on_coin_step(coin_step)
     }
 
+    /// Checks whether there are _N - f_ `Aux` messages with values in `bin_values`. If so, starts
+    /// the `Conf` round or decides.
+    fn on_bval_or_aux(&mut self) -> Result<Step<NodeUid>> {
+        if self.bin_values == BinValues::None || self.conf_values.is_some() {
+            return Ok(Step::default());
+        }
+        let (aux_count, aux_vals) = self.count_aux();
+        if aux_count < self.netinfo.num_correct() {
+            return Ok(Step::default());
+        }
+        // Execute the Common Coin schedule `false, true, get_coin(), false, true, get_coin(), ...`
+        match self.coin_state {
+            CoinState::Decided(_) => {
+                self.conf_values = Some(aux_vals);
+                self.try_update_epoch()
+            }
+            CoinState::InProgress(_) => self.send_conf(aux_vals), // Start the `Conf` message round.
+        }
+    }
+
+    /// Multicasts a `BVal(b)` message, and handles it.
+    fn send_bval(&mut self, b: bool) -> Result<Step<NodeUid>> {
+        // Record the value `b` as sent. If it was already there, don't send it again.
+        if !self.sent_bval.insert(b) {
+            return Ok(Step::default());
```

Comment:
This can happen only in case of programming error? Maybe insert a critical error here?

Reply:
No, it happens e.g. when we input `true`, and then receive the _f + 1_ st `BVal(true)`. (Or vice versa, I think.)